Modules/_testcapimodule.c

```diff
@@ -6012,6 +6012,180 @@ static struct PyModuleDef _testcapimodule = {
     NULL
 };
 
+typedef struct {
+    PyObject_HEAD
+    int value;
+} HeapCTypeObject;
+
+static struct PyMemberDef heapctype_members[] = {
+    {"value", T_INT, offsetof(HeapCTypeObject, value)},
+    {NULL} /* Sentinel */
+};
+
+static int
+heapctype_init(PyObject *self, PyObject *args, PyObject *kwargs)
+{
+    ((HeapCTypeObject *)self)->value = 10;
+    return 0;
+}
+
+static void
+heapgcctype_dealloc(HeapCTypeObject *self)
+{
+    PyTypeObject *tp = Py_TYPE(self);
+    PyObject_GC_UnTrack(self);
+    PyObject_GC_Del(self);
+    Py_DECREF(tp);
+}
+
+static PyType_Slot HeapGcCType_slots[] = {
+    {Py_tp_init, heapctype_init},
+    {Py_tp_members, heapctype_members},
+    {Py_tp_dealloc, heapgcctype_dealloc},
+    {0, 0},
+};
+
+static PyType_Spec HeapGcCType_spec = {
+    "_testcapi.HeapGcCType",
+    sizeof(HeapCTypeObject),
+    0,
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_HAVE_GC,
+    HeapGcCType_slots
+};
+
+static void
+heapctype_dealloc(HeapCTypeObject *self)
+{
+    PyTypeObject *tp = Py_TYPE(self);
+    PyObject_Del(self);
+    Py_DECREF(tp);
+}
+
+static PyType_Slot HeapCType_slots[] = {
+    {Py_tp_init, heapctype_init},
+    {Py_tp_members, heapctype_members},
+    {Py_tp_dealloc, heapctype_dealloc},
+    {0, 0},
+};
+
+static PyType_Spec HeapCType_spec = {
+    "_testcapi.HeapCType",
+    sizeof(HeapCTypeObject),
+    0,
+    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
+    HeapCType_slots
+};
+
+typedef struct {
+    HeapCTypeObject base;
+    int value2;
+} HeapCTypeSubclassObject;
+
+static int
+heapctypesubclass_init(PyObject *self, PyObject *args, PyObject *kwargs)
+{
+    /* Get HeapCTypeSubclass */
+    PyObject *m = PyState_FindModule(&_testcapimodule);
+    if (m == NULL) {
+        return -1;
+    }
+    PyTypeObject *ctypesubclass = (PyTypeObject *)PyObject_GetAttrString(m, "HeapCTypeSubclass");
+    if (ctypesubclass == NULL) {
+        return -1;
+    }
+
+    PyTypeObject *base = (PyTypeObject *)PyType_GetSlot(ctypesubclass, Py_tp_base);
+    Py_DECREF(ctypesubclass);
+    initproc base_init = PyType_GetSlot(base, Py_tp_init);
+    if (base_init(self, args, kwargs) < 0) {
```

Comment:
Is this setup to get the superclass necessary?
It should be possible to call `heapctype_init(self, args, kwargs)` directly.

Reply:
It is indeed. I actually had your suggestion initially but ultimately ended up changing it to this approach. My rationale was: To show that `HeapCTypeSubclass` in indeed a C extension subclass and that you can properly call the `tp_init` slot by pulling its base out. It seemed closer to what a normal `__init__` would do by calling `super().__init__`.

That being said, I don't have any strong opinions on either. I'll keep this as is but feel free to add one more comment to change to to call `heapctype_init` and I will change it this time!

Reply:
There are some subtle issues in doing dynamic `super()`-like calls from C code, but using `PyState_FindModule`/`PyObject_GetAttrString` is especially fragile.
If the elaborate code manages to set `base_init` to anything except `heapctype_init`, it's a bug. It's not the bug this is supposed to be testing; and it's a bug that's likely to segfault.

I'm actually trying to provide a better way to do this without `PyState_FindModule` ([PEP 573](https://www.python.org/dev/peps/pep-0573/)), but so far, using `heapctype_init` is, IMO, best.